### PR TITLE
[dom-gpu] Tool for generating Easyconfig diffs

### PR DESCRIPTION
**Pull request by vkarak**
This is a small tool I've written that generates diffs between two easyconfig specs. The problem I usually face is that since EasyBuild's mentality is to have an easyconfig for each spec (version, variant etc.) for a package, it is difficult to see what actually changes from, for example, reframe 2.3 to 2.4 or from CrayGNU 19.06 to CrayGNU 19.10. That was the motivation for this tool. It is essentially a wrapper to the system's diff command but with some EasyBuild bindings:

```
usage: ebdiff.py [-h] [-r PATH] [--num-matches NUM_MATCHES]
                 [DIFF_OPTS] PACKAGE SPEC1 SPEC2

positional arguments:
  DIFF_OPTS
  PACKAGE
  SPEC1
  SPEC2

optional arguments:
  -h, --help            show this help message and exit
  -r PATH, --robot-path PATH
                        Override EasyBuild's robot path
  --num-matches NUM_MATCHES
                        Number of relevant easyconfigs to print in case of
                        unknown specs
```

It recognizes EasyBuild's robot paths so you don't have to put full easyconfig names. It figures them out automatically. If it can't find a spec, it will give you a list of candidate easyconfigs that you might want to select sorted in order of relevance. Finally, since it wraps `diff`, you can pass it any argument that `diff` accepts and you can also customize the diff command itself through the `EBDIFF_CMD` environment variable. Finally, the tool uses Spack syntax for specifying the specs. Here are some usage examples:

- Comparing reframe versions (no toolchain info)

```
$ ./tools/ebdiff.py reframe 2.3 3.0-dev4
```
```diff
Found matching spec for reframe@2.3: /users/karakasv/Devel/production/easybuild/easyconfigs/r/reframe/reframe-2.3.eb
Found matching spec for reframe@3.0-dev4: /users/karakasv/Devel/production/easybuild/easyconfigs/r/reframe/reframe-3.0-dev4.eb
--- /users/karakasv/Devel/production/easybuild/easyconfigs/r/reframe/reframe-2.3.eb	2017-06-26 15:40:57.828202000 +0200
+++ /users/karakasv/Devel/production/easybuild/easyconfigs/r/reframe/reframe-3.0-dev4.eb	2020-04-30 16:32:03.089139000 +0200
@@ -1,25 +1,45 @@
-# @author: gppezzi
+# @author: gppezzi, vkarak, manitart

-easyblock = "Tarball"
+easyblock = 'Tarball'

 name = 'reframe'
-version = '2.3'
+version = '3.0-dev4'

-homepage = 'https://madra.cscs.ch/scs/reframe/'
-description = "A regression framework for writing portable tests for HPC systems."
+homepage = 'https://github.com/eth-cscs/reframe'
+description = 'A regression framework for writing portable tests for HPC systems.'

-toolchain = {'name': 'dummy', 'version': 'dummy'}
+toolchain = SYSTEM

-sources = [ ('archive.tar.gz?ref=v%(version)s', "tar xfz %s" )]
-source_urls = ['https://madra.cscs.ch/scs/reframe/repository/']
+sources = [
+    {
+        'filename': 'v%(version)s.tar.gz',
+        'source_urls': ['https://github.com/eth-cscs/reframe/archive/']
+    },
+    {
+        'filename': '/apps/common/easybuild/sources/%(nameletterlower)s/'
+                    '%(name)s/reframe-tests-%(version)s.tar.gz'
+    }
+]

-dependencies = [('Python-bare', '3.5.2')]
+builddependencies = [('pip', '20.0.2', '-py3')]

 keepsymlinks = True
+postinstallcmds = [
+    'cp %(installdir)s/config/cscs.py %(installdir)s/reframe/settings.py',
+    'cp -r %(builddir)s/reframe-tests-%(version)s/checks %(installdir)s/cscs-checks/private',
+    'mv %(installdir)s/cscs-checks %(installdir)s/checks',
+    'PYTHONUSERBASE=%(installdir)s/installed_by_pip pip install --upgrade '
+    '--user -r %(builddir)s/reframe-%(version)s/requirements.txt pygelf',
+]

-sanity_check_paths={
-    'files': ['reframe.py', 'bin/reframe'],
-    'dirs': ['bin', 'checks', 'unittests'],
+modextrapaths = {
+    'PYTHONPATH': 'installed_by_pip/lib/python3.6/site-packages'
 }

+sanity_check_paths = {
+    'files': ['reframe.py', 'bin/reframe', 'config/cscs.py'],
+    'dirs':  ['bin', 'checks', 'checks/private', 'unittests'],
+}
+
+sanity_check_commands = ['reframe -V', 'python3 -m pytest --version']
 moduleclass = 'devel'
```

* Comparing GROMACS (toolchains + suffixes). Note that suffixes in EasyBuild include the dash:

```
$ ./tools/ebdiff.py GROMACS 2019.2^CrayGNU@19.09 2019.2^CrayGNU@19.09^-cuda-10.1
```

```diff
Found matching spec for GROMACS@2019.2^CrayGNU@19.09: /users/karakasv/Devel/production/easybuild/easyconfigs/g/GROMACS/GROMACS-2019.2-CrayGNU-19.09.eb
Found matching spec for GROMACS@2019.2^CrayGNU@19.09^-cuda-10.1: /users/karakasv/Devel/production/easybuild/easyconfigs/g/GROMACS/GROMACS-2019.2-CrayGNU-19.09-cuda-10.1.eb
--- /users/karakasv/Devel/production/easybuild/easyconfigs/g/GROMACS/GROMACS-2019.2-CrayGNU-19.09.eb	2019-09-20 10:57:27.075689000 +0200
+++ /users/karakasv/Devel/production/easybuild/easyconfigs/g/GROMACS/GROMACS-2019.2-CrayGNU-19.09-cuda-10.1.eb	2019-10-02 15:37:23.497550000 +0200
@@ -28,7 +28,7 @@
 source_urls = ['ftp://ftp.gromacs.org/pub/gromacs/']
 sources = [SOURCELOWER_TAR_GZ]

-local_common_options = '-DCMAKE_BUILD_TYPE=Release -DGMX_BUILD_OWN_FFTW=ON -DGMX_OPENMP=ON -DGMX_GPU=OFF -DGMX_PREFER_STATIC_LIBS=ON -DGMX_SIMD=AVX2_256'
+local_common_options = '-DCMAKE_BUILD_TYPE=Release -DGPU_DEPLOYMENT_KIT_ROOT_DIR=$CUDATOOLKIT_HOME -DGMX_BUILD_OWN_FFTW=ON -DGMX_OPENMP=ON -DGMX_GPU=ON -DGMX_PREFER_STATIC_LIBS=ON -DGMX_SIMD=AVX2_256'
 configopts = ['%s -DGMX_MPI=OFF ' % local_common_options,
               '%s -DGMX_MPI=ON ' % local_common_options]

@@ -37,9 +37,11 @@
 # CMake dependency with dummy toolchain
 builddependencies = [
     ('CMake', '3.14.1', '', True),
+    ('cudatoolkit/10.1.105_3.27-7.0.1.0_8.1__ga311ce7', EXTERNAL_MODULE),
 ]

 dependencies = [
+    ('craype-accel-nvidia60', EXTERNAL_MODULE),
     ('cray-libsci/19.06.1', EXTERNAL_MODULE)
 ]

```

* Unknown specs

```
$ ./tools/ebdiff.py GROMACS 2019.2^CrayGNU@19.09 2019.2^CrayGNU@19.09^-cuda-11.1
Found matching spec for GROMACS@2019.2^CrayGNU@19.09: /users/karakasv/Devel/production/easybuild/easyconfigs/g/GROMACS/GROMACS-2019.2-CrayGNU-19.09.eb
ERROR: unknown spec: GROMACS@2019.2^CrayGNU@19.09^-cuda-11.1
The following relevant easyconfigs were found:
    - GROMACS-2019.2-CrayGNU-19.09-cuda-10.1.eb
    - GROMACS-2019.2-CrayGNU-19.10-cuda-10.1.eb
    - GROMACS-2019.2-CrayGNU-19.06-cuda-10.1.eb
    - GROMACS-2019.1-CrayGNU-19.09-cuda-10.1.eb
    - GROMACS-2019.1-CrayGNU-19.10-cuda-10.1.eb
    - GROMACS-2019.1-CrayGNU-19.06-cuda-10.1.eb
    - GROMACS-2018.6-CrayGNU-19.09-cuda-10.1.eb
    - GROMACS-2019.1-CrayGNU-19.03-cuda-10.0.eb
    - GROMACS-2018.6-CrayGNU-19.10-cuda-10.1.eb
    - GROMACS-2018.6-CrayGNU-19.06-cuda-10.1.eb
```

### CODE REVIEW

File: tools/ebdiff.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+              file=sys.stderr)
+        print_specs(matches, file=sys.stderr)
+
+    if ec2 is None:
+        print(f'ERROR: unknown spec: {options.package}@{options.spec2}',
+              file=sys.stderr)
+        print_specs(matches, file=sys.stderr)
+
+    if not do_diff:
+        sys.exit(1)
```
Comment (vkarak) [inline tools/ebdiff.py:102]: I should do exit(2) everywhere to match `diff`'s behaviour.

**Comment by lucamar** [comment]
retest this please

File: tools/ebdiff.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        print(f'    - {m}', *args, **kwargs)
+
+
+class Spec:
+    '''Wrapper for package specs.
+
+    The format is the following
+    {pkg_name}@{pkg_version}^{tc_name}@{tc_version}^{suffix}
+
+    Example: GROMACS@2019.2^CrayGNU@19.09^cuda-10.1
```
Comment (vkarak) [inline tools/ebdiff.py:41]:
```suggestion
    Example: GROMACS@2019.2^CrayGNU@19.09^-cuda-10.1
```

**Comment by boegel** [comment]
@vkarak EasyBuild already has something like this built in, which is slightly superior since it can do a 1-to-many diffs (comparing the given easyconfigs with N>1 different already existing easyconfigs), see `--preview-pr` (https://easybuild.readthedocs.io/en/latest/Integration_with_GitHub.html#previewing-easyconfig-pull-requests-preview-pr)

**Comment by victorusu** [comment]
@boegel, but the EB counterpart does not have the nice cli syntax... 😄

**Comment by lucamar** [comment]
@victorusu I tried to use the tool following the example provided with `GROMACS` but it failed: now I committed the suggestion of @vkarak to fix the `versionsuffix`.

**Comment by vkarak** [comment]
@lucamar, for the suffix you have to put the `-`. I've done this so as to be fully conformant with EB, but I can always insert it. Let me know what serves us best.

**Pull request merged**